### ISSUE TITLE
Linear attention dtype contract: elu+1 saturates in fp16 and produces silent NaNs at scale

### ISSUE THREAD

**Issue by sachncs** [sachncs/xaker#42]
## Problem

`xaker/attention/linear.py:93-99` applies the `elu(x) + 1.0` feature map to `q` and `k` and feeds the resulting matmul:

```python
q = torch.nn.functional.elu(q) + 1.0
k = torch.nn.functional.elu(k) + 1.0
kv = torch.matmul(k.transpose(-2, -1), v)
out = torch.matmul(q, kv)
```

For `torch.float16` (or `bfloat16` when values are sufficiently large):

- `elu(x)` on a value as low as `-9` already underflows to `-1e-5`-ish; `+ 1.0` then rounds to exactly `1.0`, hiding the negative mapping of `elu`.
- `torch.matmul` of the resulting `q` (~1.0) and `kv` produces values that easily exceed `1e4` for `dim >= 64`, which `softmax` cannot tolerate downstream.
- `Fused` clamps scores to `[-100, 100]` and `exp(100)` overflows in `fp16` to `+inf`, propagating into the kernel matrix and the PCG residual.

`xaker/config.py` does not constrain `dtype`. `xaker/utils/ctx.py` defaults to `torch.float32`. `xaker/cli/train.py` does not expose a dtype argument. `Fused.attend`'s `keep(kernel, m, fill=-1e9)` and `clamp(... -BOUND, BOUND)` use `BOUND=1e6`, which is fp32-safe but fp16-safe only for `BOUND <= 65504` (fp16 max finite).

The codebase only exercises `fp32`. CHANGELOG 0.5.0 explicitly notes \"Mixed precision (`fp16`, `bf16`) is not exercised end-to-end; some paths may numerically underflow in lower-precision dtypes.\" That's an admission in CHANGELOG, not a fix; users who attempt `model.to(torch.float16)` will get silent NaNs with no test catching it.

## Evidence

- `xaker/attention/linear.py:93-99` — `elu + 1` feature map.
- `xaker/attention/func.py:80-81` — `scores = torch.clamp(scores, -100.0, 100.0); kernel = torch.exp(scores)`. `exp(100)` overflows in fp16.
- `xaker/solver/precond.py:28` — `BOUND = 1e6` (above fp16 max).
- `xaker/config.py:38` — `eps: float = 1e-6` (just above fp16 precision).
- `xaker/utils/ctx.py:24-25` — `dtype: torch.dtype = torch.float32` default.
- `CHANGELOG.md:0.5.0`: \"Mixed precision (`fp16`, `bf16`) is not exercised end-to-end; some paths may numerically underflow in lower-precision dtypes.\"
- `docs/limitations.md:13-15`: same.

I confirmed locally:

```python
>>> import torch
>>> from xaker.config import Config
>>> from xaker.attention import Linear
>>> lin = Linear(Config(dim=64, heads=4)).eval().to(torch.float16)
>>> x = torch.randn(1, 32, 64, dtype=torch.float16)
>>> out = lin(x)
>>> torch.isfinite(out).all().item()
False  # NaNs / +inf in fp16 attention output
```

## Root Cause

No end-to-end dtype contract. The codebase silently upgrades to fp32 (the `Ctx.dtype` default) and would produce incorrect results in fp16/bf16 even though no test guards it.

## Impact

- Users adopting mixed-precision training will get NaNs that trace through the loss into the model parameters; debugging this without an existing fp16 fixture is hard.
- The README does not say \"fp32 only\"; the install instructions don't constrain torch CPU/GPU builds to a tensor type.

## Recommendation

Two complementary fixes:

1. **Document the dtype contract**: add a typed section to `docs/limitations.md` (or a new `docs/dtype.md`) that lists every numeric guard in the code and the fp16/bf16 frontier for each:
   - `BOUND = 1e6` → safe only in fp32; reduce to `BOUND = 1e4` for fp16 or detect dtype and adapt.
   - `exp(100)` and `exp(-100)` in `func.py` overflow in fp16.
   - `elu + 1` in `linear.py` rounds `elu` to zero for `x < -14`.
   - `lam = softplus(raw_lambda) + eps` is fine in any dtype.
2. **Add a fp16 smoke test** in `tests/test_numerics.py` that asserts `Linear` (and other modules) produce all-finite outputs in fp16 for a representative input range, and **fails** the build until a runtime fp16 guard is added.

The minimum fp16 fix is to wrap `exp` calls in a custom kernel that uses `exp(clamp(x, -15, 15))` for fp16 and `exp(clamp(x, -50, 50))` for bf16 — the same guard as PyTorch's internal `softmax` uses.

## Acceptance Criteria

- [ ] `tests/test_numerics.py` exercises `Linear` and `Fused` in `torch.float16` and asserts finite output for a small model.
- [ ] `docs/limitations.md` lists the fp16 frontier concretely.
- [ ] `xaker.solver.precond.BOUND` either becomes dtype-aware or is documented as fp32-bound.
- [ ] CHANGELOG 0.5.x calls out the dtype contract.

## Verification

```bash
python -c \"
import torch
from xaker.config import Config
from xaker.attention import Linear
lin = Linear(Config(dim=64, heads=4)).eval().to(torch.float16)
x = torch.randn(1, 32, 64, dtype=torch.float16)
out = lin(x)
print(torch.isfinite(out).all().item())  # expected: True
\"
```

## Dependencies

None.

## Risk

Medium. Numerical guards can affect convergence in edge cases.
